### PR TITLE
add Build::apple_deployment_target function

### PR DESCRIPTION
**Pull request by Be-ing**
flag_if_supported doesn't work for this because try_get_compiler adds the Apple deployment target argument before flags_supported is processed, so users can only append a second -mmacosx-version-min flag via flag_if_supported. To work around this, add a function specifically for this purpose.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -3321,7 +3333,7 @@ impl Build {
         Ok(ret)
     }
 
-    fn apple_deployment_version(
+    fn get_apple_deployment_target(
```
Comment (NobodyXu) [inline src/lib.rs:3336]: Can we change the return type to `Cow<'_, str>` to avoid cloning `str`?

**Review by NobodyXu — CHANGES_REQUESTED** [review]

**Comment by NobodyXu** [comment]
This seems like related to #902

**Comment by NobodyXu** [comment]
cc @Be-ing if you don't mind it, I can update the PR for you.

**Comment by Be-ing** [comment]
I did some more testing, and the existing `flag_if_supported` API does work. I am not sure exactly how the CLI arguments are handled by clang. When `-mmacosx-version-min` is passed twice (first time automatically by cc, second time by user-specified `flag_if_supported`), it seems to either prefer the second argument or the higher specified version. It might still be a good idea to add this API though because that behavior of clang is not documented and I don't know if it can be relied on long term. Thoughts?

**Comment by NobodyXu** [comment]
IMO it's better to add this API, not just due to the order of cmdline arg, but cc-rs seems to have different behavior based on apple deployment target version and build target.

**Comment by Be-ing** [comment]
> cc-rs seems to have different behavior based on apple deployment target version and build target

Yes, but that can be addressed using the existing `flag_if_supported` API, so should we add a new API?

**Comment by NobodyXu** [comment]
If that can be reliably solved by `flag_if_supported()` on all compilers, then it is indeed unnecessary.

But I'm not sure if it would work on gcc or msvc.

File: src/lib.rs [hunk 2]
```diff
@@ -3426,15 +3439,18 @@ impl Build {
                 }),
 
             AppleOs::Ios => deployment_from_env("IPHONEOS_DEPLOYMENT_TARGET")
+                .or_else(|| self.apple_deployment_target.as_ref().map(|a| a.to_string()))
                 .or_else(|| rustc_provided_target(rustc, target))
                 .map(maybe_cpp_version_baseline)
                 .unwrap_or_else(|| OLD_IOS_MINIMUM_VERSION.into()),
 
-            AppleOs::WatchOs => deployment_from_env("WATCHOS_DEPLOYMENT_TARGET")
+            AppleOs::WatchOs => deployment_from_env("_DEPLOYMENT_TARGET")
```
Comment (BlackHoleFox) [inline src/lib.rs:3447]: This looks like an error maybe?

Comment (Be-ing) [reply src/lib.rs:3447]: whoops, yeah that was an accident

**Comment by BlackHoleFox** [comment]
@Be-ing On the contrary I think this should still be added. `MACOSX_DEPLOYMENT_TARGET` only works if you either set it outside of the process in a Makefile, cargo env config, etc or use `std::env::set_var` at runtime inside `build.rs`. 

That last option is the only programmatic way to do it and manipulating the process' environment variables is an uh, [dicey area](github.com/rust-lang/rust/issues/90308). Having a proper library API to control this for each target would be nice imo.

**Comment by Be-ing** [comment]
Now that the old behavior of defaulting to the SDK's default deployment target has been restored, what is the need for this API? The default is a recent deployment target, so old code builds fine. A crate specifying a lower minimum than the default wouldn't really change anything.

**Comment by BlackHoleFox** [comment]
> A crate specifying a lower minimum than the default wouldn't really change anything.

What do you mean?

**Comment by Be-ing** [comment]
If a crate specifies a lower deployment target than the default, it's effectively superfluous because it would have built anyway. Only if *every* crate in the dependency graph that uses cc specifies the same lower deployment target will the resulting binary actually be usable on the lower deployment target. That's best achieved via the environment variable because that effects every crate that uses cc.

**Comment by BlackHoleFox** [comment]
Seems like a reasonable take, the well-stated counter-argument is appreciated :)

**Pull request closed without merging**